### PR TITLE
[h2olog] a new feature to declare @appdata in D script files to filter application data

### PR DESCRIPTION
**Pull request by gfx**
This PR refactors h2olog's `gen_raw_tracer.py` to recognize `/* @appdata */` annotations for probe args (see the diff in `h2o-probes.d` for details).

<del>This PR does not change any behavior in h2olog. This is my whole plan for the h2olog application-data filtering:</del>

This PR does include the following changes:

1. introduce `@appdata` annotation syntax
2. add `@appdata` to quicly-probes.d in h2o/quicly (https://github.com/h2o/quicly/pull/446)
3. add option `-a` to gather appdata, allowing h2olog users to opt-in appdata fields

FWIW I've rewritten the D script parser from the regexp-based parser to the top-down parser because it's too difficult to add new syntaxes in the regexp-based parser.

## Annotation Syntax

```c
/* @appdata
{
    "probe1": ["arg1", "arg2"],
    "probe2": ["arg3"]
}
*/
```

The `@appdata` section must be placed before the `provider` block and it can only include `@appdata { ...JSON... }`.

### CODE REVIEW

**Comment by gfx** [comment]
Set to `draft` since CI for DTrace failed by this PR.

**Comment by gfx** [comment]
Fixed in the upstream https://github.com/h2o/h2o/pull/2608 by adding `num-threads: 1`, meaning this PR might not fix the problem. Closing.

**Review by kazuho — COMMENTED** [review]
Thank you for working on this.

I hadn't assumed that we'd have to have our own parser. I appreciate that you have written one. That said, I'm not sure if we want to maintain that. I know that you're very good at these stuff, but the question is if the complexity is worth the cost, especially when the risk of misdetecting `@appdata` is going to cause privacy concerns.

How about changing the design to have a section that lists arguments of all the probes that carry application data, possibly using a well-known structured foremat? Like:
```
/*@appdata
{
  "packet_decrypted": ["bytes"]
}
*/
```

The benefits are:
* There's no need to have a complex parsing logic. All we have to do is search for a JSON that exist adjacent to a known marker.
* We can be more error-prone by detecting inconsistencies. We'd call it a compile error if the said JSON object does not exist. We'd do the same if a probe or an argument designated by the JSON elements are not found.

**Comment by gfx** [comment]
Thank you for your opinion. 

## re rewriting the parser:

I'd like to explain my rationale for rewriting it. Indeed, I've always wanted to rewrite the parser because it's already too complex to maintain. I like RegExp so much that I always make a RegExp-based parser first for any text processing, but I think the current one is too complicated for RegExp. In addition, a RegExp-based parser is hard to debug because we cannot see its internal state. 

## re avoiding a new structured format 

Agreed. JSON seems much clearer than this PR's (CSV), especially if an annotation has a single item, and also agreed to be more error-prone. Detecting inconsistencies is a must.

**Comment by kazuho** [comment]
@gfx I think I might paraphrase my point. The concern I have is that this PR is trying to build a privacy feature in a way that is prone to errors. It uses a hand-written parser without any test, and the syntax itself is also error prone.

How are we going to guarantee that the parser recognizes the annotation correctly? How are we going to guarantee that the user does not mistype the name of the annotation (e.g. misspell `appdata` as `apdata`)?

That is why I'm suggesting using separate structure for naming the privacy concerns (i.e. argnames of the probes), knowing that we cannot use the best notation anyways (that marks each argument directly).

**Comment by gfx** [comment]
I see. My understanding is that:

* the parser implementation details don't matter
* the annotation syntax doesn't matter; any explicit syntax could lead to privacy problems if one doesn't add the data attributes

So I think the only way to protect application data from exposure is to manage an allow list that has non-application data args, for example, annotating args with `@nonappdata` as the inverse of `@appdata`. With the allow-list, payloads (i.e. `char*`, `uint8_t*`, and `void*`) are not exposed by default. It's fail-safe and not error-prone, so we can focus on developer-experience rather than error-proneness on developing h2olog.

**Comment by kazuho** [comment]
> So I think the only way to protect application data from exposure is to manage an allow list that has non-application data args, for example, annotating args with @nonappdata as the inverse of @appdata.

While that is a cautious approach, I'm not sure if requiring essentially everything to be tagged is worth the complexity. I think we'd rather want to have a block list. That's what we have agreed.

IMO, the question is how we could implement a block list that we can trust.

**Comment by gfx** [comment]
Talked with @kazuho elsewhere and I  agree with https://github.com/h2o/h2o/pull/2604#pullrequestreview-603580218  :

* Place a single JSON section like `/* @appdata { ... } */` per d-script file
* We'll maintain a block list instead of an allow list since allow lists will hurt productivity and developer experience

**Comment by gfx** [comment]
It's ready to review!

File: src/h2olog/misc/gen_raw_tracer.py [hunk 1]
```diff
@@ -20,14 +20,35 @@
 # FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
 # IN THE SOFTWARE.
 
-# usage: gen-quic-bpf.py h2o_dir output_file
+# gen_raw_tracer.py - Generate an h2olog tracer to emit USDT probes by parsing D script files.
+#
+# The gen_raw_tracer.py can parse @appdata annotations in D script files, for example:
```
Comment (kazuho) [inline src/h2olog/misc/gen_raw_tracer.py:25]:
```suggestion
# gen_raw_tracer.py parses D script files accompanied by @appdata annotation, for example:
```

**Review by kazuho — COMMENTED** [review]
Thank you for working on this.

I think that `@appdata` should be a mandatory element of the probe definition files that we process. Our conclusion was that the feature has to be fail safe, and there's no reason to not require `@appdata` be defined in any of the probe definition files that we provide.

**Comment by gfx** [comment]
Thank you for reviewing my PR. 

I thought I do other TODOs separately, but finally I'd like to do them in this PR in order to keep fail-safety.

* [x] the appdata section is now required
* [x] mark appdata to all the appdata fields (quicly part is https://github.com/h2o/quicly/pull/446)
* [x] `h2olog -a` is implemented to show appdata fields

FIWI https://github.com/h2o/quicly/pull/446 is a blocker for this PR because `@appdata` section is required for quicly-probes.d.

File: h2o-probes.d [hunk 2]
```diff
@@ -22,6 +22,17 @@
 
 struct st_h2o_tunnel_t;
 
+/* @appdata
+{
+    "receive_request_header": ["value"],
+    "send_response_header": ["value"],
```
Comment (kazuho) [inline h2o-probes.d:28]: What about header field names? I think we need to discuss.

File: h2o-probes.d [hunk 3]
```diff
@@ -22,6 +22,17 @@
...
+    "h3_packet_receive": ["bytes"],
```
Comment (kazuho) [inline h2o-probes.d:29]: I do not think we need this, as it is the bytes sent over the wire.

File: h2o-probes.d [hunk 4]
```diff
@@ -22,6 +22,17 @@
...
+    "h3_frame_receive": ["bytes"],
+    "tunnel_on_read": ["bytes"],
+    "tunnel_write": ["bytes"]
```
Comment (kazuho) [inline h2o-probes.d:32]: Yes, we need these three because they contain application data in cleartext.

File: src/h2olog/misc/gen_raw_tracer.py [hunk 5]
```diff
@@ -733,15 +730,21 @@ def generate_cplusplus(context, output_file):
...
+    sys.exit(1)
+
 
 def main():
-  try:
-    (_, h2o_dir, output_file) = sys.argv
-  except:
-    print("usage: %s h2o_dir output_file" % sys.argv[0])
-    sys.exit(1)
+  output_file = sys.argv[1]
```
Comment (syohex) [inline src/h2olog/misc/gen_raw_tracer.py:723]: Python raises `IndexError` exception if `len(sys.argv) < 2`. I suppose you should check `sys.argv` length  before accessing list or wrap `try-catch` block

```python
if len(sys.argv) < 3:
    usage()

output_file, *d_files = sys.argv[1:]
```

**Review by kazuho — COMMENTED** [review]
Thank you for your continued work. Below are the reviews for appdata in h2o-probes.d.

File: h2o-probes.d (hunk 2 above)
Comment (janaiyengar) [reply h2o-probes.d:28]: Given that the method itself is in the `value` field, I don't think there's a privacy issue with letting the request header name be logged, but I also don't see the point in recording the `name`. I'd be fine with also tagging `name` here.

File: h2o-probes.d (hunk 3 above)
Comment (janaiyengar) [reply h2o-probes.d:29]: Agree with @kazuho.

**Review by janaiyengar — COMMENTED** [review]
Thank you for the work, @gfx !

File: h2o-probes.d (hunk 2 above)
Comment (gfx) [reply h2o-probes.d:28]: :ok: It makes sense. We can see them with the `-a` option in debugging, anyway.

File: src/h2olog/misc/gen_raw_tracer.py (hunk 5 above)
Comment (gfx) [reply src/h2olog/misc/gen_raw_tracer.py]: Oops. Thanks for your suggestion. Applied in 7f3668e4d3a599809172c94d7255de79bed219f6

**Comment by gfx** [comment]
Merged master to pick the quicly changes https://github.com/h2o/quicly/pull/446 via https://github.com/h2o/h2o/pull/2637

Now it's ready to merge.

File: src/h2olog/h2olog.cc [hunk 6]
```diff
@@ -232,11 +233,12 @@ int main(int argc, char **argv)
     int debug = 0;
     bool preserve_root = false;
     bool list_usdts = false;
+    bool appdata = false;
```
Comment (kazuho) [inline src/h2olog/h2olog.cc:236]:
```suggestion
    bool include_appdata = false;
```

Boolean variables above use verb + object, it'd be good to follow the convention.

File: src/h2olog/h2olog.h [hunk 7]
```diff
@@ -75,6 +75,12 @@ class h2o_tracer
         uint64_t num_events;
         uint64_t num_lost;
     } stats_;
+
+    /**
+     * Whether or not to include application data.
+     */
+    bool appdata_ = false;
```
Comment (kazuho) [inline src/h2olog/h2olog.h:82]: Here, and ...

File: src/h2olog/h2olog.h [hunk 8]
```diff
@@ -101,9 +107,10 @@ class h2o_tracer
     /**
      * Performs post-construction initialization common to all the tracers.
      */
-    void init(FILE *fp)
+    void init(FILE *fp, bool appdata)
```
Comment (kazuho) [inline src/h2olog/h2olog.h:110]: here also.

**Review by kazuho — APPROVED** [review]
Thank you for all the work. LGTM with a nitpick.

**Pull request merged**